daemon: add multiple snap sideload to API

MiguelPires opened the pull request:
Add support for sideloading multiple local snaps.

bboozzoo commented on daemon/api_sideload_n_try.go:
```diff
@@ -123,38 +125,119 @@ func sideloadOrTrySnap(c *Command, body io.ReadCloser, boundary string) Response
+	return chg, nil
+}
+
+func readSideInfo(st *state.State, form *Form, tempPath string, origPath string) (*snap.SideInfo, *apiError) {
```
we don't really need Form at this level. There can be:
```
type sideloadFlags struct {
    SnapModeFlags snapstate.Flags
    DangerousOK bool
}

func readSideInfo(st *state.State, tempPath, origPath string, flags sideloadFlags) (*snap.SideInfo, *apiError) {
```

bboozzoo commented on daemon/api_sideload_n_try.go:
```diff
@@ -123,38 +125,119 @@ func sideloadOrTrySnap(c *Command, body io.ReadCloser, boundary string) Response
+func readSideInfo(st *state.State, form *Form, tempPath string, origPath string) (*snap.SideInfo, *apiError) {
 	var sideInfo *snap.SideInfo
 
+	dangerousOK := isTrue(form, "dangerous")
 	if !dangerousOK {
```
```suggestion
	if !flags.DangerousOK {
```

and the `!isTrue(form, "devmode")` further down can be replaced with `flags.SnapModeFlags & snapstate.DevMode == 0` (or could be a sideloadFlags level thing itself

bboozzoo commented on daemon/api_sideload_n_try.go:
```diff
@@ -123,38 +125,119 @@ func sideloadOrTrySnap(c *Command, body io.ReadCloser, boundary string) Response
+	return chg, nil
+}
+
+func sideloadSnap(st *state.State, form *Form, origPath, tempPath string, flags snapstate.Flags) (*state.Change, *apiError) {
```
```suggestion
func sideloadSnap(st *state.State, sideloadSnap uploadedSnap, flags sideloadFlags) (*state.Change, *apiError) {
```

bboozzoo commented on daemon/api_sideload_n_try.go:
```diff
@@ -123,38 +125,119 @@ func sideloadOrTrySnap(c *Command, body io.ReadCloser, boundary string) Response
+
+func sideloadSnap(st *state.State, form *Form, origPath, tempPath string, flags snapstate.Flags) (*state.Change, *apiError) {
 	var instanceName string
 	if len(form.Values["name"]) > 0 {
```
this can then check `sideloadSnap.InstanceName != ""` and form is not needed anymore

bboozzoo commented on daemon/api_sideload_n_try.go:
```diff
@@ -75,24 +77,25 @@ func (f *Form) RemoveAllExcept(paths []string) {
-	if len(f.Values["snap-path"]) > 0 {
-		srcFilename = f.Values["snap-path"][0]
+	if len(srcFilenames) == 1 && len(f.Values["snap-path"]) > 0 {
+		srcFilenames[0] = f.Values["snap-path"][0]
```
I wonder if we should actively complain if this is the case.

MiguelPires replied:
You mean return 400 if there's more than one snap and a snap-path? Hm not sure, don't really have a strong opinion. The original path was optional and only used in the error/change summary so there's not much impact either way I think

bboozzoo reviewed (commented):
Some comments

bboozzoo commented on daemon/api_sideload_n_try_test.go:
```diff
@@ -645,6 +645,128 @@ func (s *sideloadSuite) TestCleanUpUnusedTempSnapFiles(c *check.C) {
+
+	var data map[string][]string
+	c.Assert(chg.Get("api-data", &data), check.IsNil)
+	c.Check(data["snap-names"], testutil.DeepUnsortedMatches, snaps)
```
The names come in order of appearance so you can use DeepEquals here.

bboozzoo commented on daemon/api_sideload_n_try_test.go:
```diff
@@ -645,6 +645,128 @@ func (s *sideloadSuite) TestCleanUpUnusedTempSnapFiles(c *check.C) {
+	})
+	defer readRest()
+
+	body := "----hello--\r\n" +
+		"Content-Disposition: form-data; name=\"devmode\"\r\n" +
+		"\r\n" +
+		"true\r\n" +
+		"----hello--\r\n"
```
It'd be nice to have a test with dangerous too.

We have this pattern that may you'll find useful:

```
func (s *sideloadSuite) testSideloadManySnaps(c *check.C, flagsBody string, expectedFlags snapstate.Flags) {

}

func (s *sideloadSuite) TestSideloadManySnapsDevmode(c *check.C) {
	body := "----hello--\r\n" +
		"Content-Disposition: form-data; name=\"devmode\"\r\n" +
		"\r\n" +
		"true\r\n" +
		"----hello--\r\n"
	expectedFlags := snapstate.Flags{RemoveSnapPath: true, DevMode: true}
	s.testSideloadManySnaps(c,  *check.C, body, expectedFlags)
}

func (s *sideloadSuite) TestSideloadManySnapsDangerous(c *check.C) {
	body := "----hello--\r\n" +
		"Content-Disposition: form-data; name=\"dangerous\"\r\n" +
		"\r\n" +
		"true\r\n" +
		"----hello--\r\n"
	expectedFlags := snapstate.Flags{RemoveSnapPath: true}
	s.testSideloadManySnaps(c,  *check.C, body, expectedFlags)
}

func (s *sideloadSuite) TestSideloadManySnapsAsserted(c *check.C) {
	// TODO setup assertions
	body := "----hello--\r\n"
	expectedFlags := snapstate.Flags{RemoveSnapPath: true}
	s.testSideloadManySnaps(c,  *check.C, body, expectedFlags)
}
```

pedronis replied:
yes, would be good

bboozzoo commented on daemon/api_sideload_n_try_test.go:
```diff
@@ -645,6 +645,128 @@ func (s *sideloadSuite) TestCleanUpUnusedTempSnapFiles(c *check.C) {
+	c.Check(apiErr.JSON().Status, check.Equals, 400)
+	c.Check(apiErr.Message, check.Equals, `cannot read snap file: expected`)
+}
+
```
We probably need another test where dangerous or devmode are not set and one of the snaps isn't asserted.

bboozzoo reviewed (commented):
Thanks for the updates. I proposed some little tweaks and test improvements. Feel free to ping me when you need help with mocking assertions.

pedronis commented on daemon/api_sideload_n_try.go:
```diff
@@ -123,97 +153,156 @@ func sideloadOrTrySnap(c *Command, body io.ReadCloser, boundary string) Response
+		return nil, errToResponse(err, tempPaths, InternalError, "cannot install snap files: %v")
+	}
+
+	msg := fmt.Sprintf(i18n.G("Install snaps %q from files %q"), names, origPaths)
```
%q here produces `["name1", "name2"...]`  which is not what we use. These should use instead %s and strutil.Quoted. You can see examples in api_snaps.go I think.

pedronis reviewed (commented):
overall looks good. I agree with @bboozzoo about testing a bit more. Some other small remarks.

MiguelPires commented on daemon/api_sideload_n_try.go:
```diff
@@ -123,97 +153,156 @@ func sideloadOrTrySnap(c *Command, body io.ReadCloser, boundary string) Response
+		userID = user.ID
+	}
+
+	tss, err := snapstateInstallPathMany(context.TODO(), st, sideInfos, tempPaths, userID, &flags.Flags)
```
Yes, the disk space checks use it to calculate the space used by prerequisites

pedronis replied:
we probably need to think more about this, I forgot that installSize calls the store but that doesn't make complete consitent sense to do in the preparation phase of a sometimes local-only operation

pedronis replied:
as disk-space checking is behind a feature flag  and the issue is really in checkDiskSpace/installSize we agreed to fix this in a follow-up

MiguelPires commented:
Thanks for the feedback. I addressed your comments in the last two commits, with the exception of the open question about disk space checks. That will require more thought/discussion but it can be dealt with in a follow-up PR

bboozzoo reviewed (approved):
LGTM, thank you!

mardy commented on daemon/api_sideload_n_try.go:
```diff
@@ -123,97 +153,156 @@ func sideloadOrTrySnap(c *Command, body io.ReadCloser, boundary string) Response
 		return errRsp
 	}
 
+	systemRestartImmediate := isTrue(form, "system-restart-immediate")
+	if systemRestartImmediate {
+		chg.Set("system-restart-immediate", true)
+	}
```
Unless `"system-restart-immediate" == false` has some special meaning, I'd just write:
```suggestion
	chg.Set("system-restart-immediate", isTrue(form, "system-restart-immediate"))
```

MiguelPires replied:
Changed. Thanks!

mardy reviewed (approved):
LGTM, though it's hard to review since the original functions are very long.

I have just nitpicks, feel free to ignore.

pedronis commented on daemon/api_sideload_n_try_test.go:
```diff
@@ -645,6 +648,283 @@ func (s *sideloadSuite) TestCleanUpUnusedTempSnapFiles(c *check.C) {
+	rsp := s.errorReq(c, req, nil)
+
+	c.Assert(rsp.Status, check.Equals, 400)
+	// tries to read file to get sideinfo
```
maybe:

// get as far as reading file to get side info

would be clearer

pedronis reviewed (approved):
thanks
